### ISSUE TITLE
ShouldBeUnique-Jobs store their locks inside the app-file-cache when database-connection is used, causing problems with no-unique-criteria-jobs on detached environments

### ISSUE THREAD

**Issue by Husky110** [laravel/framework#41464]
- Laravel Version: 9.1.1 (artisan said 9.4.3) but can be replicated with 8.6.11 aswell - I did not test more downward.
- PHP Version: 8.1.3

### Description:
Okay, this was a bit of an odyssey to figure out, but I will try my best to describe the problem - or better: the possible multiple problems.  
- Given you have an environment where the database-server is somehow detached from the laravel-application (in my case I am running a docker-compose-environment where one container acts as dispatcher and other containers run as workers doing the actual heavy-duty).
- Now you implement any given job and implement the "ShouldBeUnique"-Interface without any matching uniqueId-, uniqueVia- or uniqueFor-logic in there - just the plain job with the "implements ShouldBeUnique"-interface and nothing more.
- Last you set your QUEUE_CONNECTION-value to "database". 

Now dispatch that job (via `MyJob::dispatch()`) and Laravel dispatches the job nicely into the jobs-table. Now let's say, you don't actually process this job (keeping it inside the table) and dispatch it again. The expected behaviour would be, that this job is not beeing dispatched again - in fact when you try to dispatch it, you get an object of type "Illuminate\Foundation\Bus\PendingDispatch" with the job as object, but "null" on the job-property of that object.  
If that in itself is already a problem is debateable, cause it would be nice to have a message there like "Can't dispatch job, since it already exists" or something in that matter on the PendingDispatch-Object. But that might be just me "wishing" for something that is impossible. Back to the real problem.  
We have our job sitting inside the jobs-table. Now let's say something happens to the job-entry inside the table. A careless delete on the jobs-table or the job actually beeing processed by another server with a `$this->delete();` at the end (yep - that causes the main-problem aswell). What would you expect Laravel to do, if someone calls `MyJob::dispatch()` on the ShouldBeUnique-Job? And here do the problems begin... 

I would expect the following workflow to happen: 
- Laravel checks the jobs-table to see if the given job is already dispatched
- In the stated situation above, Laravel would see no instance of the job beeing inside the jobs-table and dispatch a new instance

What's actually happening...  
For that we have to take a step back.  
Laravel actually stores the lock for the ShouldBeUnique-jobs inside it's application cache. That means that the following workflow is happening:
- Laravel checks it's application-cache for a lock on the given job
- Laravel finds a lock-entry in it's cache
- Laravel does NOT dispatch the given job

So far this might read like a user-problem, but in my oppinion it is not. I am explicitly stating that I want to use the database as my QUEUE_CONNECTION (env-name). So I am stating that ALL queue-logic-data (like locks, timeouts, etc.) shall be stored inside my database. Appearently Laravel ignores that, which then results in jobs not beeing dispatched. And like stated above - if the job runs on a worker, which is another server or docker-container and the job calls the `$this->delete();` at the end - the job can never be dispatched again by the dispatching server (container). This behaviour actually makes queues not useable for anyone running detached servers.

### Steps To Reproduce:
(are given in the description)

**Comment by driesvints** [comment]
> I am explicitly stating that I want to use the database as my QUEUE_CONNECTION (env-name). So I am stating that ALL queue-logic-data (like locks, timeouts, etc.) shall be stored inside my database. 

Thank you for your report but this isn't a bug. Like the documentation says, all unique jobs, regardless of which queue connection is used require a cache store that support locks in order to function. If you want this to behave differently, you're welcome to attempt a PR.

<img width="725" alt="Screenshot 2022-03-14 at 16 01 38" src="https://user-images.githubusercontent.com/594614/158200190-0eb948fb-5e80-442f-a9ce-0fcd6c75e7af.png">

**Comment by Husky110** [comment]
@driesvints - Okay, I see your point there. But I hope agree that the documentation is really missing the "regardless of which queue connection is used"-part in there. Before filing this issue, I talked to other developers about my problem and noone was able to say "Hey - did you check the cache-driver?". So am I (or the other devs) to blame to not knowing this, or the docs to be more specific? Plus - the documentation states nowhere that the application dispatching the job and the application doing the work have to be the same application. So every developer beeing new to the topic could easily make the assumption I made and go even further in writing 2 complete different apps without knowing the desaster he/she is approaching.  
I think this one needs more discussion on how to approach this because as of now I am sure that I am not the only one having a problem with the current state. Most ppl I talked to about this actually were like "Wait, what!? Seriously!?". I would kindly ask you to reopen the issue to resolve it properly and I see two ways for this:
- Update the docs so they are more specific and warn about the problem of detaching applications which have the `ShouldBeUnique`-Interface
- Actually update Laravel so it actually checks it's locks on the queue it is using.

To me the latter would make more sense, since I don't see why subsystem A has to rely on subsystem B when the idea of Laravel is to combine multiple subsystems into a easy-to-use framework. Just closing it and be like "Well - it is how it is. Deal with it!" feels a bit like sweeping the problem under the carpet and hoping noone finds it there. Until the next person stumbles.
